lib/bluetooth_low_energy/aws_ble_gatt.c

```diff
@@ -740,302 +555,159 @@ static void vIndicationSentCb( uint16_t usConnId,
+                break;
 
-            xSemaphoreGive( ( SemaphoreHandle_t ) &xBTInterface.xThreadSafetyMutex );
+            case eBTDbCharacteristic:
+                xStatus = xBTInterface.pxGattServerInterface->pxAddCharacteristic( xBTInterface.ucServerIf,
+                                                                                   pxService->pusHandlesBuffer[ 0 ],
+                                                                                   &pxService->pxBLEAttributes[ usAttributes ].xCharacteristic.xUuid,
+                                                                                   pxService->pxBLEAttributes[ usAttributes ].xCharacteristic.xProperties,
+                                                                                   pxService->pxBLEAttributes[ usAttributes ].xCharacteristic.xPermissions );
+                break;
+
+            default:
+                xStatus = eBTStatusFail;
+        }
+
+        if( xStatus == eBTStatusSuccess )
+        {
+            xEventGroupWaitBits( ( EventGroupHandle_t ) &xBTInterface.xWaitOperationComplete,
+                                 1 << eBLEHALEventAttributeAddedCb,
+                                 pdTRUE,
+                                 pdTRUE,
+                                 portMAX_DELAY );
         }
         else
         {
-            xStatus = eBTStatusFail;
+            break;
         }
     }
-    else
-    {
-        xStatus = eBTStatusParamInvalid;
-    }
 
     return xStatus;
 }
 
 /*-----------------------------------------------------------*/
 
-BTStatus_t BLE_StartService( BLEService_t * pxService,
-                             BLEServiceStartedCallback_t pxCallback )
+BTStatus_t BLE_CreateService( BLEService_t * pxService,
+                              BLEAttributeEventCallback_t pxEventsCallbacks[] )
 {
-    BTStatus_t xStatus = eBTStatusSuccess;
+    BTStatus_t xStatus = eBTStatusParamInvalid;
 
+    /* Create all attributes. */
     if( pxService != NULL )
     {
-        xBTInterface.pxPendingServiceEvent = ( void * ) pxCallback;
-        xStatus = xBTInterface.pxGattServerInterface->pxStartService( xBTInterface.ucServerIf,
-                                                                      pxService->xAttributeData.xHandle,
-                                                                      BTTransportLe );
+        memset( pxService->pusHandlesBuffer, 0, pxService->xNumberOfAttributes );
+        xStatus = prvAddServiceToList( pxService, pxEventsCallbacks );
     }
-    else
+
+    /* After all attributes have been create successfully, the service is added to the list. */
+    if( xStatus == eBTStatusSuccess )
     {
-        xStatus = eBTStatusFail;
+        if( xSemaphoreTake( ( SemaphoreHandle_t ) &xBTInterface.xThreadSafetyMutex, portMAX_DELAY ) == pdPASS )
+        {
+            xStatus = prvCreateAttributes( pxService );
+            xSemaphoreGive( ( SemaphoreHandle_t ) &xBTInterface.xThreadSafetyMutex );
+        }
```

Comment:
use assert if it fails?
But we can do this when we rename the API, so you can use the new assert style, and logging too... just let's remember

Reply:
ok, I ll add it